examples/hitl/rearrange_v2/app_state_start_session.py

```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Meta Platforms, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import List, Optional
+
+from app_data import AppData
+from app_state_base import AppStateBase
+from app_states import (
+    create_app_state_cancel_session,
+    create_app_state_load_episode,
+)
+from session import Session
+
+from habitat_hitl.app_states.app_service import AppService
+
+
+class AppStateStartSession(AppStateBase):
+    def __init__(self, app_service: AppService, app_data: AppData):
+        super().__init__(app_service, app_data)
+        self._new_session: Optional[Session] = None
+        self._save_keyframes = False
+
+    def get_next_state(self) -> Optional[AppStateBase]:
+        episode_ids = self._try_get_episode_ids()
+        if episode_ids is not None:
+            # Start the session.
+            self._new_session = Session(
+                self._app_service.config,
+                list(episode_ids),
+                dict(self._app_data.connected_users),
+            )
+
+            if self._cancel:
+                return create_app_state_cancel_session(
+                    self._app_service,
+                    self._app_data,
+                    self._new_session,
+                    "User disconnected",
+                )
+            else:
+                return create_app_state_load_episode(
+                    self._app_service, self._app_data, self._new_session
+                )
+        else:
+            # Create partial session record for data collection.
+            self._new_session = Session(
+                self._app_service.config,
+                [],
+                dict(self._app_data.connected_users),
+            )
+            return create_app_state_cancel_session(
+                self._app_service,
+                self._app_data,
+                self._new_session,
+                "Invalid session",
+            )
```

Comment:
Error reporting is limited.

In most cases, users are kicked when an error occurs (e.g. they are AFK), which triggers cancellation of the session.
This is a consequence of the older HITL systems which did not have any state management.

This is sloppy, but "good enough".

With this new state machine, it would be relatively easy to propagate errors. However, that would require some refactoring ensure that HITL applications can redefine the behavior when an invalid state is reached (instead of kicking).